contracts/governance/src/governance.rs

```diff
@@ -10,81 +10,121 @@ use alloc::vec;
 use canonical_derive::Canon;
 use dusk_bls12_381::BlsScalar;
-use dusk_jubjub::GENERATOR_EXTENDED;
-use dusk_pki::PublicKey;
-use dusk_schnorr::Signature;
+use dusk_bls12_381_sign::{PublicKey as BlsPublicKey, Signature};
+use dusk_bytes::Serializable;
 
-use crate::collection::Collection;
+use crate::collection::{Map, Set};
 use crate::*;
 
-#[derive(Debug, Default, Clone, Canon)]
+#[derive(Debug, Clone, Default, Canon)]
 pub struct GovernanceContract {
-    pub(crate) seeds: Collection<BlsScalar, ()>,
-    pub(crate) balances: Collection<PublicKey, u64>,
-    pub(crate) whitelist: Collection<PublicKey, ()>,
+    pub(crate) seeds: Set<BlsScalar>,
+    pub(crate) balances: Map<PublicKey, u64>,
+    pub(crate) whitelist: Set<PublicKey>,
     pub(crate) paused: bool,
     pub(crate) total_supply: u64,
+    // we use BlsPublicKey or dusk_bls12_381_sign::PublicKey and not a
+    // dusk_pki::PublicKey because of our verification method
+    //
+    // They need to be public so they can be accessed from the host environment
+    // (rusk). Once contract deployment stragery will be implemented, this will
+    // change.
+    pub broker: BlsPublicKey,
+    pub authority: BlsPublicKey,
 }
 
+/// Use `GovernanceContract::default()` for instance.
 impl GovernanceContract {
-    /// Authority of the contract
-    ///
-    /// Will have to be defined in the constant space so the bytecode of the
-    /// contract will be changed as the authority does
-    pub const AUTHORITY: PublicKey =
-        PublicKey::from_raw_unchecked(GENERATOR_EXTENDED);
-
-    fn validate_seed(
+    fn assert_seed(
         &mut self,
         arguments: Vec<BlsScalar>,
         signature: Signature,
     ) -> Result<(), Error> {
         let seed = arguments[0];
 
-        if self.seeds.get(&seed)?.is_some() {
+        if self.seeds.contains(&seed) {
             return Err(Error::SeedAlreadyUsed);
         }
 
         #[cfg(target_arch = "wasm32")]
-        if !rusk_abi::verify_schnorr_sign(
+        if !rusk_abi::verify_bls_sign(
             signature,
-            Self::AUTHORITY,
-            rusk_abi::poseidon_hash(arguments),
+            self.authority,
+            rusk_abi::poseidon_hash(arguments).to_bytes().to_vec(),
         ) {
             return Err(Error::InvalidSignature);
         }
 
         #[cfg(not(target_arch = "wasm32"))]
-        if !signature
-            .verify(&Self::AUTHORITY, dusk_poseidon::sponge::hash(&arguments))
+        if self
+            .authority
+            .verify(
+                &signature,
+                &dusk_poseidon::sponge::hash(&arguments).to_bytes(),
+            )
+            .is_ok()
         {
             return Err(Error::InvalidSignature);
         }
 
-        self.seeds.insert(seed, ())?;
+        self.seeds.insert(seed);
 
         Ok(())
     }
 
-    fn check_pause(&self) -> Result<(), Error> {
-        (!self.paused).then_some(()).ok_or(Error::ContractIsPaused)
+    fn assert_pause(&self) -> Result<(), Error> {
```

Comment:
The function name is a bit misleading.
To me `assert_pause` means that we need to ensure that the contract is paused, otherwise it has to return an Err

But the current behavior is the opposite